Added solver: 6pt rel pose + focal (equal for both cameras), RFC for F estimation

kocurvik opened the pull request:
As promised in #89 I have implemented these two features so far:
* 6pt relative pose solver with two unknown, but equal focal lengths
  - generated using autogen by V. Larsson et al.
  - LO for robust estimation
  - adds Danilevsky charpoly from AM (from autogen) to sturm.h
  - changes max depth of Sturm's bisection method to 300 - **this could affect performance of other solvers**, but it is necessary for the implemented solver
  - new struct "CalibratedCameraPose" holding a camera and pose - could be refactored in future
  - added basic python interface for Camera
  - includes python call for solver and robust estimation
  - includes benchmark - this is somewhat hacky as it randomly generates configurations until all points are in front of both cameras
* RFC checking for imaginary focals as proposed in: [https://arxiv.org/abs/2311.16304](https://arxiv.org/abs/2311.16304)
  - adds RFC for F estimation via inheritance from FundamentalEstimator (perhaps using C++ templates would be better)
  - turned on via RansacOpts rfc, requires rfc_pp1, rfc_pp2 in opts as well - maybe not the best solution

I realized that the focal refinement from the paper has quite a lot of redundant calculations in the original C++ implementation. I will add it later to this PR or a new one. I will also probably add 6pt solver for a calibrated camera and camera with unknown focal based on feedback to this PR.

Benchmark results:
```
                    Solver    Solutions        Valid     GT found      Runtime
------------------------------------------------------------------------------
                       p3p      1.72738          100          100   443.613 ns
                  p3p_ding      1.72738          100          100   347.152 ns
                      gp3p        3.827          100          100   1.86929 us
                     gp4ps            1          100        99.98   2.37865 us
               gp4ps(Deg.)       2.2956          100        99.99    506.14 ns
                      p4pf       0.9999        99.99        99.43   3.24378 us
                    p2p2pl        6.708       99.836         99.4   38.6516 us
                      p6lp       4.4418          100        99.99   2.07397 us
               p5lp_radial       2.2801      99.9965       99.998   1.53775 us
                    p2p1ll            4       99.995        99.97    1.8013 us
                    p1p2ll       4.4764          100        99.99   1.96123 us
                      p3ll       4.5554      99.9978        99.94   2.03711 us
                      up2p            2          100          100   150.279 ns
                   up1p1ll            2          100          100   124.278 ns
                     ugp2p            2          100          100   151.165 ns
                    ugp3ps            1          100          100   508.583 ns
                   up1p2pl       3.2768          100          100   2.63793 us
                     up4pl        4.526      99.9646        99.87    1.8941 us
                    ugp4pl       4.4624      99.9462        99.85   1.88397 us
             RelUpright3pt       2.5022       99.988        99.99     452.7 ns
          GenRelUpright4pt       4.6296      99.8769        99.53   1.92623 us
                    Rel8pt            1          100          100    2.8513 us
           Rel8pt(100 pts)            1          100          100   9.50269 us
                    Rel5pt       2.2442      98.7033        98.62   7.79852 us
               RelFocal6pt       1.8469      97.8938         97.7   46.4695 us
       RelUprightPlanar2pt       1.8064          100          100     228.2 ns
       RelUprightPlanar3pt            1          100          100    808.63 ns
               GenRel5p1pt       1.1579      98.8082        47.76   7.85224 us
                 GenRel6pt        2.489      99.9598         99.8   1.00564 ms
             Homography4pt            1          100          100   792.432 ns
          Homography4pt(C)       0.9924          100        99.24   737.882 ns
```

The RelFocal6pt outputs correct result in less than 98 percent of cases. This is possibly due to the bad numerics when calculating eigenvalues.

I also tested it with posebench (see [PR in posebench](https://github.com/vlarsson/posebench/pull/1)):

```
Dataset: scannet1500_sift
6pt pose + shared f (poselib): AUC5= 8.23, AUC10=18.47, AUC20=31.29, f AUC5=14.92, f AUC10=24.44, f AUC20=35.93, avg_rt= 52.7ms, med_rt= 52.7ms
Dataset: scannet1500_spsg
6pt pose + shared f (poselib): AUC5=10.76, AUC10=24.20, AUC20=41.50, f AUC5=12.97, f AUC10=23.53, f AUC20=37.19, avg_rt= 48.8ms, med_rt= 48.9ms
```

I also tested RFC with posebench:
```
Dataset: megadepth1500_sift
F (poselib)  : AUC5= 9.11, AUC10=14.87, AUC20=23.21, avg_rt= 18.5ms, med_rt= 17.2ms
F+RFC (poselib): AUC5= 8.08, AUC10=13.52, AUC20=21.02, avg_rt= 11.2ms, med_rt= 10.0ms
Dataset: megadepth1500_spsg
F (poselib)  : AUC5=49.82, AUC10=65.01, AUC20=77.15, avg_rt= 14.0ms, med_rt= 12.2ms
F+RFC (poselib): AUC5=49.32, AUC10=64.66, AUC20=77.33, avg_rt= 12.5ms, med_rt= 10.7ms
Dataset: megadepth1500_splg
F (poselib)  : AUC5=51.08, AUC10=66.31, AUC20=78.34, avg_rt= 11.1ms, med_rt=  9.7ms
F+RFC (poselib): AUC5=50.76, AUC10=65.91, AUC20=78.08, avg_rt=  9.8ms, med_rt=  8.3ms
Dataset: scannet1500_sift
F (poselib)  : AUC5= 5.74, AUC10=13.24, AUC20=23.09, avg_rt=  3.8ms, med_rt=  4.0ms
F+RFC (poselib): AUC5= 6.05, AUC10=14.27, AUC20=25.00, avg_rt=  3.4ms, med_rt=  3.1ms
Dataset: scannet1500_spsg
F (poselib)  : AUC5= 8.32, AUC10=19.42, AUC20=34.72, avg_rt= 13.2ms, med_rt= 12.2ms
F+RFC (poselib): AUC5= 8.59, AUC10=19.84, AUC20=35.01, avg_rt= 10.2ms, med_rt=  9.2ms
Dataset: imc_british_museum
F (poselib)  : AUC5=17.80, AUC10=30.16, AUC20=45.14, avg_rt=  6.7ms, med_rt=  5.4ms
F+RFC (poselib): AUC5=17.21, AUC10=29.36, AUC20=44.18, avg_rt=  6.0ms, med_rt=  5.0ms
Dataset: imc_london_bridge
F (poselib)  : AUC5=39.28, AUC10=51.25, AUC20=61.20, avg_rt=  5.8ms, med_rt=  5.0ms
F+RFC (poselib): AUC5=38.81, AUC10=51.08, AUC20=61.73, avg_rt=  5.1ms, med_rt=  4.1ms
Dataset: imc_piazza_san_marco
F (poselib)  : AUC5= 9.23, AUC10=18.06, AUC20=32.18, avg_rt=  9.0ms, med_rt=  5.9ms
F+RFC (poselib): AUC5= 9.30, AUC10=18.46, AUC20=32.80, avg_rt=  7.9ms, med_rt=  5.0ms
Dataset: imc_florence_cathedral_side
F (poselib)  : AUC5=56.10, AUC10=70.81, AUC20=81.72, avg_rt= 11.1ms, med_rt=  8.0ms
F+RFC (poselib): AUC5=55.82, AUC10=70.46, AUC20=81.49, avg_rt=  9.8ms, med_rt=  7.0ms
Dataset: imc_milan_cathedral
F (poselib)  : AUC5=34.65, AUC10=51.17, AUC20=66.50, avg_rt= 10.2ms, med_rt=  7.0ms
F+RFC (poselib): AUC5=33.99, AUC10=50.49, AUC20=65.94, avg_rt=  9.0ms, med_rt=  6.0ms
Dataset: imc_sagrada_familia
F (poselib)  : AUC5=41.10, AUC10=56.22, AUC20=68.96, avg_rt= 13.0ms, med_rt=  6.2ms
F+RFC (poselib): AUC5=41.26, AUC10=56.41, AUC20=69.24, avg_rt= 11.9ms, med_rt=  5.7ms
Dataset: imc_lincoln_memorial_statue
F (poselib)  : AUC5=36.87, AUC10=47.87, AUC20=57.47, avg_rt=  5.1ms, med_rt=  4.0ms
F+RFC (poselib): AUC5=36.35, AUC10=47.25, AUC20=56.74, avg_rt=  4.6ms, med_rt=  4.0ms
Dataset: imc_mount_rushmore
F (poselib)  : AUC5=30.39, AUC10=42.63, AUC20=55.65, avg_rt= 11.0ms, med_rt=  6.1ms
F+RFC (poselib): AUC5=29.81, AUC10=42.16, AUC20=55.25, avg_rt= 10.0ms, med_rt=  5.5ms
Dataset: imc_st_pauls_cathedral
F (poselib)  : AUC5=33.82, AUC10=50.05, AUC20=65.24, avg_rt=  7.8ms, med_rt=  6.0ms
F+RFC (poselib): AUC5=34.27, AUC10=50.62, AUC20=65.62, avg_rt=  6.8ms, med_rt=  5.0ms

All of the above combined:
F (poselib)  : AUC5=30.24, AUC10=42.65, AUC20=55.04, avg_rt=  8.9ms, med_rt=  6.7ms
F+RFC (poselib): AUC5=29.96, AUC10=42.45, AUC20=54.94, avg_rt=  7.5ms, med_rt=  5.5ms
```

It is ~10-20 % faster with generally comparable results.

pablospe commented:
Check C++ formatting:
https://github.com/PoseLib/PoseLib/actions/runs/7876199471/job/21496196436?pr=92

You can run this line at the top-level in the repo:
https://github.com/PoseLib/PoseLib/actions/runs/7876199471/workflow?pr=92#L31

pablospe commented:
Check the CI pipeline, the compilation failed:
https://github.com/PoseLib/PoseLib/actions/runs/7886661232/job/21521586955?pr=92

vlarsson commented on PoseLib/camera_pose.h:
```diff
@@ -66,6 +67,18 @@ struct alignas(32) CameraPose {
 
 typedef std::vector<CameraPose> CameraPoseVector;
+
+struct CalibratedCameraPose {
```
This struct is a bit weird and this is partly because we use CameraPose for both absolute and relative poses. For relative pose, it would make more sense to have two cameras in the struct. Also, when I think about a calibrated camera, I usually mean [R t], i.e what CameraPose already is :) I am not entirely sure what would be a better name though. 

Maybe use 
Image { pose, camera } ?
ImagePair { relpose, camera1, camera2 } ?

This would be sort of consistent with COLMAP terminology well

kocurvik replied:
I can rename the struct to Image and create ImagePair. If that would be the case then what should be the output of robust_shared_focal_relative_pose? The Image struct would contain all the information (using Image is also easier internally in LO), but since both cameras are calibrated it would also make sense to use ImagePair with the two cameras being the same.

kocurvik replied:
I added Image and ImagePair and now the estimator outputs ImagePair

kocurvik commented on PoseLib/types.h:
```diff
@@ -47,6 +47,9 @@ struct RansacOptions {
     bool progressive_sampling = false;
     size_t max_prosac_iterations = 100000;
+    bool rfc = false;
+    Eigen::Vector2d rfc_pp1 = Eigen::Vector2d::Zero();
```
I implemented it this way for ease of use so the users do not have to transform F back to original principal points. However, in my own implementation I also assumed that the points have been centered. I will change this back.

kocurvik replied:
Now the F estimation w/ RFC assumes that the points have been centered.

kocurvik commented on PoseLib/robust/estimators/relative_pose.h:
```diff
@@ -133,6 +162,13 @@ class FundamentalEstimator {
     std::vector<size_t> sample;
 };
 
+class FundamentalEstimatorRFC : public FundamentalEstimator {
```
I will change this. My motivation was to avoid an if statement in F estimation without RFC, but it probably does not affect the performance in a significant way anyways.

kocurvik replied:
Changed

kocurvik commented on PoseLib/misc/sturm.h:
```diff
@@ -205,7 +206,7 @@ void ridders_method_newton(const double *fvec, double a, double b, double *roots
 void isolate_roots(const double *fvec, const double *svec, double a, double b, int sa, int sb, double *roots,
                    int &n_roots, double tol, int depth) {
-    if (depth > 30)
+    if (depth > 300)
```
This are the output results of benchmark for the original max depth = 30 on my PC:
```
                    Solver    Solutions        Valid     GT found      Runtime
------------------------------------------------------------------------------
                       p3p      1.72738          100          100   448.883 ns
                  p3p_ding      1.72738          100          100   345.232 ns
                      gp3p        3.827          100          100   1.86583 us
                     gp4ps            1          100        99.98   2.37792 us
               gp4ps(Deg.)       2.2982          100          100    463.62 ns
                      p4pf       0.9998        99.98        99.41   3.25747 us
                    p2p2pl        6.708       99.836         99.4   38.5888 us
                      p6lp       4.4418          100        99.99   2.13638 us
               p5lp_radial       2.2801      99.9965       99.998   1.53835 us
                    p2p1ll            4       99.995        99.97   1.82671 us
                    p1p2ll       4.4758          100        99.97   2.01345 us
                      p3ll       4.5552      99.9978        99.94   2.14563 us
                      up2p            2          100          100   152.367 ns
                   up1p1ll            2          100          100   125.685 ns
                     ugp2p            2          100          100   151.634 ns
                    ugp3ps            1          100          100   506.074 ns
                   up1p2pl       3.2768          100          100   2.71015 us
                     up4pl        4.526      99.9646        99.87   1.95649 us
                    ugp4pl       4.4624      99.9462        99.85   1.89346 us
             RelUpright3pt       2.5022       99.988        99.99    470.82 ns
          GenRelUpright4pt       4.6296      99.8769        99.53   1.96094 us
                    Rel8pt            1          100          100   2.92895 us
           Rel8pt(100 pts)            1          100          100   9.58045 us
                    Rel5pt       2.2439      98.7032        98.61   7.76357 us
               RelFocal6pt        1.706      98.2181         91.5   45.1464 us
       RelUprightPlanar2pt       1.8064          100          100    225.37 ns
       RelUprightPlanar3pt            1          100          100    811.18 ns
               GenRel5p1pt       1.1577       98.808        47.75   7.86209 us
                 GenRel6pt        2.489      99.9598         99.8   1.00626 ms
             Homography4pt            1          100          100   734.116 ns
          Homography4pt(C)       0.9924          100        99.24   723.234 ns
```

There seems to be no performance degradation for Rel5pt when comparing with the output posted in the PR's description (for max depth = 300 on the same machine).

sarlinpe replied:
Unfortunately I don't have an easy way to reproduce this. Using a tolerance makes sense but IIUC this does not prevent deep recursions for degenerate cases in which two roots are very close to each other. These cases could maybe be caught earlier?

vlarsson replied:
For the bisection method we use here, I think either using recursion, or maintaining your own explicit stack is necessary, as it cannot be rewritten to tail recursion (we need to traverse both legs). But I guess other methods could be used. The easiest to implement would probably be to just build the companion matrix and use the Eigen-solver in Eigen. We could have this as an option (as a compile flag) to replace`bisect_sturm` with this.

kocurvik commented on PoseLib/robust.cc:
```diff
@@ -230,6 +230,58 @@ RansacStats estimate_relative_pose(const std::vector<Point2D> &points2D_1, const
     return stats;
 }
 
+RansacStats estimate_focal_relative_pose(const std::vector<Point2D> &points2D_1, const std::vector<Point2D> &points2D_2,
```
Changed all the relevant names.

kocurvik commented on PoseLib/solvers/relpose_6pt_focal.cc:
```diff
@@ -0,0 +1,1147 @@
+
+        double focal = std::sqrt(1.0 / sols(2, i).real());
+
+        Eigen::VectorXd F_vector = N.col(0) + sols(0, i).real() * N.col(1) + sols(1, i).real() * N.col(2);
```
Changed to Eigen::Vector<double, 9>

pablospe commented on PoseLib/camera_pose.h:
```diff
@@ -68,17 +68,30 @@ struct alignas(32) CameraPose {
-    CalibratedCameraPose() : pose(CameraPose()), camera(Camera()) {}
-    CalibratedCameraPose(CameraPose pose, Camera camera) : pose(pose), camera(camera) {}
+    // Constructors (Defaults to identity camera and pose)
+    Image() : pose(CameraPose()), camera(Camera()) {}
```
Image inside `camera_pose.h` might be confusing. Also, has an image a pose? Image is a 2D thing here? and pose a 3D one. CalibratedCameraPose might be actually better name than `Image` name. Maybe `Keyframe`? Perhaps I am missing something here.

kocurvik replied:
I assume @vlarsson meant Image as it is [used in COLMAP](https://github.com/colmap/colmap/blob/f21d604b968bfaf47f9af4e52dbd8a405e1f8942/src/colmap/scene/image.h#L48-L51). It holds information about both pose and camera used to take the image.

I could put the Image and Image structs in a different file, or `camera_pose.h` could be renamed.

kocurvik commented on PoseLib/misc/sturm.h:
```diff
@@ -273,6 +274,47 @@ template <> inline int bisect_sturm<1>(const double *coeffs, double *roots, doub
+        for (int j = 0; j <= i; j++)
+            A.row(j) = A.row(j) + Acol(j) * vinv.transpose();
+
+        A.row(i) = Eigen::VectorXd::Zero(n);
```
Changed.

vlarsson commented on PoseLib/types.h:
```diff
@@ -47,6 +47,7 @@ struct RansacOptions {
     // If we should use PROSAC sampling. Assumes data is sorted
     bool progressive_sampling = false;
     size_t max_prosac_iterations = 100000;
+    bool rfc = false;
```
I am happy with the PR now. I will do some testing on my own but otherwise it looks good. We might want to have something more descriptive for the option name here. 

Maybe real_focal_check. We should also add a comment here explaining that it's for F estimation and requires centered principal points (and probably a reference to the paper)

kocurvik replied:
I have changed this and added a comment.

vlarsson commented on PoseLib/camera_pose.h:
```diff
@@ -66,6 +67,31 @@ struct alignas(32) CameraPose {
 
 typedef std::vector<CameraPose> CameraPoseVector;
+
+struct Image {
```
I guess Image and ImagePair also need the Eigen alignment macros as well?

kocurvik replied:
I have added the macro and alignment, but I am not sure how exactly this works.

vlarsson commented on PoseLib/camera_pose.h:
```diff
@@ -66,6 +67,31 @@ struct alignas(32) CameraPose {
+struct ImagePair {
+    // Struct simply holds information about two cameras and their relative pose
+    CameraPose pose;
+    Camera camera_1;
```
I think I prefer camera1

kocurvik commented:
I have made the requested changes. I assume that the new solver and robust estimator should be included in the README. This probably also applies to RFC and the new Image and ImagePair objects and their bindings. I can prepare the README changes as well.

kocurvik commented:
I have changed the version and README. I have added the added Image and ImagePair classes together with the description of the CameraPose. Instructions for RFC are included as a comment in the RansacStats struct shown in README. The runtime for the solver was estimated based on the relative performance w.r.t. other solvers on my machine.

kocurvik commented:
> @kocurvik any plans on integrating focals_from_F ? https://github.com/kocurvik/robust_self_calibration/blob/main/cxx/iterative_focal.cpp#L140

I am planning to implement this into poselib. I already had a preliminary version within poselib, but after some additional tests it seems that the C++ version is less stable than the Matlab implementation so it needs some tuning. I also realized that the prior for the principal point can always be assumed to be [0, 0] (as it also was in the experiments) so a lot of the expressions in the current code are redundant as they evaluate to 0.

I hope that I will have some time during the summer to put this together.

vlarsson commented:
This would be great! Something else that comes to mind would be to have good code for homography decomposition (similar to OpenCV decomposeHomography).
